libs/ktem/ktem/index/file/pipelines.py

```diff
@@ -236,226 +249,356 @@ def get_pipeline(cls, user_settings, index_settings, selected):
+
+        shutil.copy(file_path, self.FSPath / file_hash)
+        source = self.Source(
+            name=file_path.name,
+            path=file_hash,
+            size=file_path.stat().st_size,
+            user=self.user_id,  # type: ignore
+        )
+        with Session(engine) as session:
+            session.add(source)
+            session.commit()
+            file_id = source.id
 
-        # persist the files to storage
-        for path in to_index:
-            shutil.copy(path, filestorage_path / file_to_hash[path])
+        return file_id
 
-        # extract the file & prepare record info
-        file_to_source: dict = {}
-        extraction_errors = []
-        nodes = []
-        for file_path, file_hash in file_to_hash.items():
-            if str(Path(file_path).resolve()) not in to_index:
-                continue
+    def finish(self, file_id: str, file_path: Path) -> str:
+        """Finish the indexing"""
+        with Session(engine) as session:
+            stmt = select(self.Index.target_id).where(self.Index.source_id == file_id)
+            doc_ids = [_[0] for _ in session.execute(stmt)]
+            if doc_ids:
+                docs = self.DS.get(doc_ids)
+                stmt = select(self.Source).where(self.Source.id == file_id)
+                result = session.execute(stmt).first()
+                if result:
+                    item = result[0]
+                    item.text_length = sum([len(doc.text) for doc in docs])
+                    session.add(item)
+                session.commit()
+
+        return file_id
+
+    def run(self, file_path: str | Path, reindex: bool, **kwargs) -> str:
+        """Index the file and return the file id"""
+        # check for duplication
+        file_path = Path(file_path).resolve()
+        file_id = self.get_id_if_exists(file_path)
+        if file_id is not None:
+            if not reindex:
+                raise ValueError(
+                    f"File {file_path.name} already indexed. Please rerun with "
+                    "reindex=True to force reindexing."
+                )
+            else:
+                # remove the existing records
+                with Session(engine) as session:
+                    session.execute(
+                        delete(self.Source).where(self.Source.id == file_id)
+                    )
+                    vs_ids, ds_ids = [], []
+                    index = session.execute(
+                        select(self.Index).where(self.Index.source_id == file_id)
+                    ).all()
+                    for each in index:
+                        if each[0].relation_type == "vector":
+                            vs_ids.append(each[0].target_id)
+                        else:
+                            ds_ids.append(each[0].target_id)
+                        session.delete(each[0])
+                    session.commit()
+                self.VS.delete(vs_ids)
+                self.DS.delete(ds_ids)
```

Comment:
this logic should be made into separate `delete_file()` function to simplify the logic of `run()` and avoid the duplication in `stream()`

Reply:
Changed in [bfff3c9](https://github.com/Cinnamon/kotaemon/pull/81/commits/bfff3c9a5aa1be55d4b79caff046a3f5858c9d3d)